Duet AI  documentation lookup prompts for Beam ML (#30351)
apache/beam: learning/prompts/documentation-lookup/36_ai_kfp_orchestration.md

Prompt:
How can I orchestrate Apache Beam pipelines with KubeFlow?
Response:
Apache Beam is an open source, unified programming model for defining both batch and streaming data-parallel processing pipelines.Each Apache Beam pipeline is a Directed Acyclic Graph (DAG) of data processing operations constructed using Apache Beam SDK in one of the [supported languages](https://beam.apache.org/documentation/sdks/java/).
To run a pipeline you need to deploy it to one of the supported [Beam runners](https://beam.apache.org/documentation/runners/capability-matrix/).
You can use Apache Beam as an integral part of a machine learning project for data ingestion, validation, preprocessing, model validation, and inference.
A full ML workflow also contains other steps such as data exploration, feature engineering, model training, and model deployment. Furthermore, ML workflow needs to track metadata and artifacts for reproducibility and auditability. Popular tools for building and managing ML workflows include [KubeFlow](https://www.kubeflow.org/), [TFX](https://www.tensorflow.org/tfx), and [Apache Airflow](https://airflow.apache.org/).

>  `KubeFlow` is an open source ML platform that is designed to enable MLOps - a set of best practices and guiding principles aimed to streamline development and maintenance of ML systems. KubeFlow provides a set of tools for building, deploying, and managing end-to-end ML pipelines in the form of a `DAG` responsible for scheduling and running the pipeline steps and passing execution parameters, metadata, and artifacts between the steps.

Reviewer: ```suggestion
Kubeflow is an open-source ML platform tailored for MLOps, which comprises best practices and guiding principles to streamline the development and maintenance of ML workflows. Kubeflow provides a set of tools for building, deploying, and managing end-to-end ML pipelines in the form of DAGs. The DAGs orchestrate the scheduling and execution of pipeline steps and facilitate the transfer of execution parameters, metadata, and artifacts between steps.
```

> You can make Apache Beam pipeline part of a Kubeflow pipeline. In this case Apache Beam pipeline `DAG` becomes a node in the Kubeflow pipeline `DAG`.

Reviewer: ```suggestion
You can integrate an Apache Beam pipeline into a Kubeflow pipeline, where the Apache Beam pipeline DAG serves as a node within the Kubeflow pipeline DAG.
```

> In order to run an Apache Beam pipeline as part of a Kubeflow pipeline you need to:

Reviewer: ```suggestion
To run an Apache Beam pipeline as a part of a Kubeflow Pipeline (KFP), you need to:
```

1. Create KFP components by specifying interfaces and containerizing the component implementation.

> 2. Create KFP pipeline by connecting components, and defining how inputs and outputs are passed between the components.

Reviewer: ```suggestion
2. Create the KFP pipeline by connecting components and defining how inputs and outputs are exchanged between them.
```

> 3. Compile and run KFP pipeline by submitting it to a KFP client endpoint.

Reviewer: ```suggestion
3. Compile and run the KFP pipeline by submitting it to a KFP client endpoint.
```

> Following is an example of a KFP pipeline that orchestrates an Apache Beam preprocessing pipeline:

Reviewer: ```suggestion
Here is an example of a KFP pipeline orchestrating an Apache Beam preprocessing pipeline:
```

```
    kfp
    ├── pipeline.py
    ├── components
    │   └── preprocess
    │       ├── Dockerfile
    │       ├── component.yaml
    │       ├── requirements.txt
    │       └── src
    │           └── preprocess.py
    └── requirements.txt
```

> Apache Beam pipelines can be used as `KFP` components consisting of a yaml specification `component.yaml` and a python source file `preprocess.py`. The yaml file specifies the input and output arguments of the component, and the python file contains the Apache Beam pipeline code.

Reviewer: ```suggestion
Apache Beam pipelines can be used as KFP components consisting of a YAML specification `component.yaml` and a Python source file `preprocess.py`. The YAML file defines the input and output arguments of the component, while the Python file contains the Apache Beam pipeline code.
```

> Because `KFP` provides the input and output arguments as command-line arguments, an `argumentparser` is needed. Component logic and requirements are containerized and the container image is pushed to a container registry.

Reviewer: ```suggestion
Since KFP provides input and output arguments as command-line arguments, an `ArgumentParser` is required. The component logic and requirements are containerized, and the container image is pushed to a container registry.
```

> As a final step, the KFP pipeline is compiled to a json file and submitted to a KFP client endpoint in `pipeline.py`.

Reviewer: ```suggestion
As a final step, the KFP pipeline is compiled into a JSON file and submitted to a KFP client endpoint using `pipeline.py`.
```

```python

> 

Reviewer: ```suggestion
```

Compiler().compile(pipeline_func=pipeline, package_path="pipeline.json")

client = kfp.Client()
experiment = client.create_experiment("KFP orchestration example")
run_result = client.run_pipeline(
    experiment_id=experiment.id,
    job_name="KFP orchestration job",
    pipeline_package_path="pipeline.json",
    params=run_arguments)
```

> See [here](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/examples/ml-orchestration/kfp) for a complete example of orchestrating Apache Beam preprocessing pipeline with KubeFlow.

Reviewer: ```suggestion

For a complete example of orchestrating an Apache Beam preprocessing pipeline with Kubeflow, refer to the [Apache Beam GitHub repository](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/examples/ml-orchestration/kfp).
```
